### PR TITLE
netty: make GrpcHttp2ConnectionHandler able to indicate it will no longer be used

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
This adds a method on GrpcHttp2ConnectionHandler which, when called, indicates that the channel associated with the handler is no longer needed.   

Notes:

* The handler may not be on the channel, but will either a. need to be added or will never be added.
* The channel will only be "unused" on the server side.
* It is expected that after calling `notifyUnused()`, the channel will be deregistered from the loop without being properly shut down.   This allows the channel to be handed off to a Non-netty API.

### CODE REVIEW

File: netty/src/main/java/io/grpc/netty/NettyServer.java [hunk 1]
```diff
@@ -156,6 +157,24 @@ public void start(ServerListener serverListener) throws IOException {
...
+          @Override
+          public void operationComplete(ChannelFuture future) throws Exception {
+            if (!done) {
+              done = true;
+              eventLoopReferenceCounter.release();
+            }
+          }
+        }
+
+        ChannelFutureListener loopReleaser = new LoopReleaser();
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyServer.java]: Move this to just before its usage?

File: netty/src/main/java/io/grpc/netty/GrpcHttp2ConnectionHandler.java [hunk 2]
```diff
@@ -34,6 +35,8 @@ public GrpcHttp2ConnectionHandler(Http2ConnectionDecoder decoder,
     super(decoder, encoder, initialSettings);
   }
 
+  protected ChannelPromise channelUnused;
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/GrpcHttp2ConnectionHandler.java]: Place before the constructor?

Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/GrpcHttp2ConnectionHandler.java]: Could this be initialized from the constructor? Having the super-class reach in is harder to follow. It still being @Nullable is fine. I'd be fine with either two constructors (one with and one without the promise) or having the client pass in a `null` instance.

Comment (carl-mastrangelo) [reply netty/src/main/java/io/grpc/netty/GrpcHttp2ConnectionHandler.java]: Done.  When I was making this CL, I thought there was a cycle between creating the channel (for the promise) and the handler.  That only ended up in the tests, so Done.

File: netty/src/main/java/io/grpc/netty/NettyServer.java (hunk 1 above)
Comment (carl-mastrangelo) [reply netty/src/main/java/io/grpc/netty/NettyServer.java]: Done.

**Pull request merged**